### PR TITLE
Add new data transfer hive udtf, bug fix & support complex types

### PR DESCRIPTION
**Pull request by cornmonster**
Improvements:
1. Support complex types and BINARY
2. Refactor code in hive native way to avoid potential bugs
3. bug fix: DATETIME & TIMESTAMP will cause NullPointerException
4. bug fix: Partition table sometimes cause ArrayOutOfOrder Exception

### CODE REVIEW

File: odps-data-carrier/data-transfer-hive-udtf/src/main/java/com/aliyun/odps/datacarrier/transfer/OdpsDataTransferUDTF.java [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+  private String[] trimAll(String[] array) {
+    for (int i = 0; i < array.length; i++) {
+      array[i] = array[i].trim();
+    }
+    return array;
+  }
+
+  private Object getOdpsObject(ObjectInspector objectInspector, Object value, TypeInfo typeInfo) {
+    if (objectInspector.getCategory().equals(Category.PRIMITIVE)) {
```
Comment (lyman) [inline odps-data-carrier/data-transfer-hive-udtf/src/main/java/com/aliyun/odps/datacarrier/transfer/OdpsDataTransferUDTF.java]: better to replace all these if-else with a hash map

Comment (cornmonster) [inline odps-data-carrier/data-transfer-hive-udtf/src/main/java/com/aliyun/odps/datacarrier/transfer/OdpsDataTransferUDTF.java]: Replace with a Map<Category -> static class>

File: odps-data-carrier/data-transfer-hive-udtf/src/main/java/com/aliyun/odps/datacarrier/transfer/OdpsDataTransferUDTF.java [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    this.objectInspectors = args;
+    // This UDTF doesn't output anything
+    return ObjectInspectorFactory.getStandardStructObjectInspector(new ArrayList<String>(),
+        new ArrayList<ObjectInspector>());
+  }
+
+  @Override
+  public void process(Object[] args) throws HiveException {
+    try {
+      StringObjectInspector soi0 = (StringObjectInspector) objectInspectors[0];
```
Comment (cornmonster) [inline odps-data-carrier/data-transfer-hive-udtf/src/main/java/com/aliyun/odps/datacarrier/transfer/OdpsDataTransferUDTF.java]: The table name & columns & partition columns are not likely to change during the life time of the UDTF, so we can read them only once to reduce overhead

File: odps-data-carrier/hive_udtf_runner.py [hunk 3]
```diff
@@ -2,9 +2,6 @@
 import sys
```
Comment (lyman) [inline odps-data-carrier/hive_udtf_runner.py]: please add licence

File: odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/report/HiveTableDivision.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
+package com.aliyun.odps.datacarrier.odps.datacarrier.report;
```
Comment (lyman) [inline odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/report/HiveTableDivision.java]: please add licence

File: odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/report/ReportBuilder.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
+package com.aliyun.odps.datacarrier.odps.datacarrier.report;
```
Comment (lyman) [inline odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/report/ReportBuilder.java]: please add licence

File: odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/report/SummaryDivision.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package com.aliyun.odps.datacarrier.odps.datacarrier.report;
```
Comment (lyman) [inline odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/report/SummaryDivision.java]: please add licence

File: odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/report/ReportBuilder.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        .head()
+          .link().attrRel(EnumRelType.STYLESHEET)
+                 .attrType(EnumTypeContentType.TEXT_CSS)
+                 .attrHref("style.css")
+          .__()
+        .__();
+
+    Body<Html<HtmlView>> body = html.body();
+    body
+        .h1().text("ODPS compatibility report").__();
```
Comment (lyman) [inline odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/report/ReportBuilder.java]: official name should be MaxCompute

File: odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/report/HiveDatabaseDivision.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
+package com.aliyun.odps.datacarrier.odps.datacarrier.report;
```
Comment (lyman) [inline odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/report/HiveDatabaseDivision.java]: please add licence

File: odps-data-carrier/bin/hive_udtf_sql_runner.py [hunk 9]
```diff
@@ -64,19 +65,57 @@ def main(root: str, udtf_resource_path: str) -> None:
...
+                print("INFO: execute %s failed, retrying..." % file_path)
+              retry -= 1
+
+            if retry == 0:
+              print("ERROR: execute %s failed 5 times" % file_path)
 
 
 if __name__ == '__main__':
-    if len(sys.argv) != 3:
+    if len(sys.argv) != 2:
```
Comment (lyman) [inline odps-data-carrier/bin/hive_udtf_sql_runner.py]: try lib argparser to parse arguments, will make life much easier

File: odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/MetaProcessor.java [hunk 10]
```diff
@@ -297,7 +308,7 @@ private String getMultiPartitionHiveUdtfSQL(DatabaseMetaModel databaseMeta,
   public static void main(String[] args) throws Exception {
     if (args.length != 2) {
```
Comment (lyman) [inline odps-data-carrier/meta-processor/src/main/java/com/aliyun/odps/datacarrier/odps/datacarrier/MetaProcessor.java]: [nice-to-have] try common-cli to parse arguments in java

**Pull request merged**